Generate serialization for WebCore::Color

sheeparegreat opened the pull request:
#### 71513d364e3d4cb1f32a6f8fcf8b387ce57e8ac2
<pre>
Generate serialization for WebCore::Color
<a href="https://bugs.webkit.org/show_bug.cgi?id=263089">https://bugs.webkit.org/show_bug.cgi?id=263089</a>
rdar://116882741

Reviewed by Ryosuke Niwa and Alex Christensen.

* Source/WebCore/platform/graphics/Color.cpp:
(WebCore::Color::Color):
(WebCore::Color::data const):
* Source/WebCore/platform/graphics/Color.h:
(WebCore::Color::encode const): Deleted.
(WebCore::Color::decode): Deleted.
* Source/WebCore/platform/graphics/ColorSpace.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/269336@main">https://commits.webkit.org/269336@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fe166e320fbec01be3dae009b69ee933fa90ca2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22248 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22447 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23320 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24145 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20587 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22495 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26709 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22774 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/21600 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/22190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24999 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/19240 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20160 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26404 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/20257 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20390 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24268 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20821 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17714 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/20177 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24387 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2782 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20768 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/platform/graphics/ColorTypes.h:
```diff
@@ -721,6 +736,26 @@ struct RGBA {
+        return value & 0xFF;
+    }
+
     uint32_t value;
```
Let's just serialize this uint32_t instead of breaking it into pieces.  Then this will need no modification.

achristensen07 commented on Source/WebCore/platform/graphics/ColorTypes.h:
```diff
@@ -201,6 +201,16 @@ constexpr bool operator==(const ColorType& a, const ColorType& b)
 
+enum class ColorSpace : uint8_t;
+
+struct OutOfLineColorData {
```
This should be a nested struct inside ColorData.

weinig replied:
OutOfLineColorData is not part of the RGB color types family and should not be here. Please keep this isolated to WebCore::Color.

rniwa commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,51 @@ Color::Color(Color&& other)
+        };
+
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
+                 .usesFunctionSerialization = flags().contains(FlagsIncludingPrivate::UseColorFunctionSerialization), // NOLINT
```
Don't align lines like this. Always indent with 4 spaces instead.

achristensen07 replied:
Don't use NOLINT here.  Just make the indentation more like this:
return {
    something,
    somethingElse
};

rniwa commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,51 @@ Color::Color(Color&& other)
+
+    } else {
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
+                 .usesFunctionSerialization = flags().contains(FlagsIncludingPrivate::UseColorFunctionSerialization), // NOLINT
```
Ditto.

achristensen07 commented on Source/WebCore/platform/graphics/Color.h:
```diff
@@ -61,10 +61,17 @@ class Color {
         UseColorFunctionSerialization   = 1 << 1,
     };
 
+    struct ColorData {
```
ColorDataForIPC?

rniwa commented on Source/WebCore/platform/graphics/ColorTypes.h:
```diff
@@ -201,6 +201,16 @@ constexpr bool operator==(const ColorType& a, const ColorType& b)
+
+struct OutOfLineColorData {
+    ColorSpace colorSpace;
+    float c1;
```
Please spell out red, green, blue instead of c1, c2, c3.

weinig replied:
This is not rgb data, it is just generic 3 channel color, so c1, c2, c3 is the canonical naming.

weinig commented on Source/WebCore/platform/graphics/ColorTypes.h:
```diff
@@ -711,6 +721,11 @@ struct RGBA {
     {
     }
 
+    explicit RGBA(uint8_t red, uint8_t green, uint8_t blue, uint8_t alpha)
```
Why is this needed?

achristensen07 replied:
It's not if we serialize the uint32_t instead.

sheeparegreat commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,51 @@ Color::Color(Color&& other)
+        OutOfLineColorData oolcd = {
+            .colorSpace = colorSpace(),
+            .c1 = c1,
+            .c2 = c2,
```
.c3 = c3.

achristensen07 commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
+    if (colorData.usesFunctionSerialization)
+        flags.add(FlagsIncludingPrivate::UseColorFunctionSerialization);
+
+    if (std::holds_alternative<PackedColor::RGBA>(colorData.data)) {
```
This should use WTF::switchOn

achristensen07 commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
+        auto& d = std::get<PackedColor::RGBA>(colorData.data);
+        setColor(asSRGBA(d), flags);
+    } else if (std::holds_alternative<OutOfLineColorData>(colorData.data)) {
+        flags.add({ FlagsIncludingPrivate::Valid, FlagsIncludingPrivate::OutOfLine });
```
This wasn't in Color::decode

achristensen07 commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
+        flags.add(FlagsIncludingPrivate::UseColorFunctionSerialization);
+
+    if (std::holds_alternative<PackedColor::RGBA>(colorData.data)) {
+        flags.add({ FlagsIncludingPrivate::Valid });
```
This wasn't in Color::decode

achristensen07 commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
+            .alpha = alpha
+        };
+
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
```
nit: it would be more consistent if this were on the next line.

achristensen07 commented on Source/WebCore/platform/graphics/Color.cpp:
```diff
@@ -49,6 +49,54 @@ Color::Color(Color&& other)
+        };
+
+    } else {
+        return { .isSemantic = flags().contains(FlagsIncludingPrivate::Semantic),
```
ditto

achristensen07 commented on Source/WebCore/platform/graphics/Color.h:
```diff
@@ -61,10 +61,25 @@ class Color {
         UseColorFunctionSerialization   = 1 << 1,
     };
 
+    struct OutOfLineColorData {
```
This can be nested in ColorDataForIPC.

sheeparegreat committed:
Generate serialization for WebCore::Color
https://bugs.webkit.org/show_bug.cgi?id=263089
rdar://116882741

Reviewed by Ryosuke Niwa and Alex Christensen.

* Source/WebCore/platform/graphics/Color.cpp:
(WebCore::Color::Color):
(WebCore::Color::data const):
* Source/WebCore/platform/graphics/Color.h:
(WebCore::Color::encode const): Deleted.
(WebCore::Color::decode): Deleted.
* Source/WebCore/platform/graphics/ColorSpace.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/269336@main
